internal/config/cacher.go

```diff
@@ -0,0 +1,203 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"sync"
+	"time"
+)
+
+// Cacher will detect which config values are read when constructing a value.
+// Then, when further requests are made, it will return the same value as long
+// as the config values which were used don't change.
+type Cacher struct {
+	cfg View
+	m   sync.Mutex
+
+	r *rememberingView
+	v interface{}
+}
+
+// NewCacher returns a cacher which uses cfg to detect relevant changes.
+func NewCacher(cfg View) *Cacher {
+	return &Cacher{
+		cfg: cfg,
+	}
+}
+
+// Get will call f the first time it is called.  It remembers which values are
+// read from the provided View, and the value returned.  When Get is next
+// called, it will check if any of the used values have changed.  If they
+// haven't, it will return the cache'd value.  Otherwise it will call f again.
+// If f returns an error, the value will not be cached or returned.
+func (c *Cacher) Get(f func(cfg View) (interface{}, error)) (interface{}, error) {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	if c.r == nil || c.r.hasChanges() {
+		c.r = newRememberingView(c.cfg)
+		var err error
+		c.v, err = f(c.r)
+		if err != nil {
+			c.r = nil
+			c.v = nil
+			return nil, err
+		}
+	}
+
+	return c.v, nil
+}
+
+// ForceReset causes Cacher to forget the remembered value.  The next call to
+// Get will again construct a new value.
+func (c *Cacher) ForceReset() {
+	c.m.Lock()
+	defer c.m.Unlock()
+	c.r = nil
+	c.v = nil
+}
+
+type rememberingView struct {
```

Comment:
cachedView? (I assume rememberingView implies the view that was last cached? - adding a comment here would make it easier to understand)

Reply:
It's not really a cache because it's not returning any of the values it remembers.  Changed to viewChangeDetector with a comment.